java/src/jmri/jmrix/openlcb/swing/lccpro/LccProFrame.java

```diff
@@ -0,0 +1,735 @@
+import java.awt.event.*;
+import java.awt.datatransfer.Transferable;
+
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.util.ArrayList;
+
+import javax.swing.*;
+import javax.swing.event.ListSelectionEvent;
+
+import jmri.InstanceManager;
+import jmri.ShutDownManager;
+import jmri.UserPreferencesManager;
+
+import jmri.swing.ConnectionLabel;
+import jmri.swing.JTablePersistenceManager;
+import jmri.swing.RowSorterUtil;
+
+import jmri.jmrix.ActiveSystemsMenu;
+import jmri.jmrix.ConnectionConfig;
+import jmri.jmrix.ConnectionConfigManager;
+import jmri.jmrix.can.CanSystemConnectionMemo;
+import jmri.jmrix.openlcb.swing.TrafficStatusLabel;
+
+import jmri.util.HelpUtil;
+import jmri.util.WindowMenu;
+import jmri.util.datatransfer.RosterEntrySelection;
+import jmri.util.swing.JmriAbstractAction;
+import jmri.util.swing.JmriJOptionPane;
+import jmri.util.swing.JmriMouseAdapter;
+import jmri.util.swing.JmriMouseEvent;
+import jmri.util.swing.JmriMouseListener;
+import jmri.util.swing.multipane.TwoPaneTBWindow;
+
+import org.openlcb.MimicNodeStore;
+
+/**
+ * A window for LCC Network management.
+ * <p>
+ *
+ * @author Bob Jacobsen Copyright (C) 2010, 2016, 2024
+ * @author Kevin Dickerson Copyright (C) 2011
+ * @author Randall Wood Copyright (C) 2012
+ */
+public class LccProFrame extends TwoPaneTBWindow  {
+
+    static final ArrayList<LccProFrame> frameInstances = new ArrayList<>();
+    protected boolean allowQuit = true;
+    protected JmriAbstractAction newWindowAction;
+
+    CanSystemConnectionMemo memo;
+    MimicNodeStore nodestore;
+    
+    public LccProFrame(String name) {
+        this(name,
+            jmri.InstanceManager.getNullableDefault(jmri.jmrix.can.CanSystemConnectionMemo.class));
+    }
+
+    public LccProFrame(String name, CanSystemConnectionMemo memo) {
+        this(name,
+                "xml/config/parts/apps/gui3/lccpro/LccProFrameMenu.xml",
+                "xml/config/parts/apps/gui3/lccpro/LccProFrameToolBar.xml",
+                memo);
+    }
+
+    public LccProFrame(String name, String menubarFile, String toolbarFile) {
+        this(name, menubarFile, toolbarFile, jmri.InstanceManager.getNullableDefault(jmri.jmrix.can.CanSystemConnectionMemo.class));
+    }
+
+    public LccProFrame(String name, String menubarFile, String toolbarFile, CanSystemConnectionMemo memo) {
+        super(name, menubarFile, toolbarFile);
+        this.memo = memo;
+        if (memo == null) {
+            // a functional LccFrame can't be created without an LCC ConnectionConfig
+            javax.swing.JOptionPane.showMessageDialog(this, "LccPro requires a configured LCC or OpenLCB connection, will quit now",
+               "LccPro", JOptionPane.ERROR_MESSAGE);
+            // and close the program
+            // This is justified because this should never happen in a properly
+            // configured system.
+            InstanceManager.getDefault(jmri.ShutDownManager.class).shutdown();
```

Comment:
Just checking. I assume that if this happens, there is a way for the user to configure a proper connection?
I ask since this code will shutdown JMRI immediately and therefore not letting the user configure the connection.

Reply:
Unfortunately, there's no reasonable way to get back to the preferences at this point.  The context isn't available this deep inside the program.  But note the comment about "should never happen in a properly configured system".  That refers not to the user level configuration, but to how the JMRI application has been built.  There are separate checks at startup (which will open the preferences) and in the LCC/OpenLCB menus that should prevent ever getting to this place.  I'll update the comment to make this clearer.